src/Markdig/Syntax/MarkdownObjectExtensions.cs

```diff
@@ -1,131 +1,119 @@
-                if (childT != null)
-                {
-                    yield return childT;
-                }
-
-                var subContainer = child as ContainerInline;
-                if (subContainer != null)
-                {
-                    foreach (var sub in subContainer.Descendants<T>())
-                    {
-                        yield return sub;
-                    }
-                }
-
-                child = next;
-            }
-        }
-
-        /// <summary>
-        /// Iterates over the descendant elements for the specified markdown <see cref="Block" /> element and filters by the type {T}.
-        /// </summary>
-        /// <typeparam name="T">Type to use for filtering the descendants</typeparam>
-        /// <param name="block">The markdown object.</param>
-        /// <returns>
-        /// An iteration over the descendant elements
-        /// </returns>
-        public static IEnumerable<T> Descendants<T>(this ContainerBlock block) where T : Block
-        {
-            foreach (var subBlock in block)
-            {
-                var subBlockT = subBlock as T;
-                if (subBlockT != null)
-                {
-                    yield return subBlockT;
-                }
-
-                var subBlockContainer = subBlock as ContainerBlock;
-                if (subBlockContainer != null)
-                {
-                    foreach (var sub in subBlockContainer.Descendants<T>())
-                    {
-                        yield return sub;
-                    }
-                }
-            }
-        }
-    }
+// Copyright (c) Alexandre Mutel. All rights reserved.
+// This file is licensed under the BSD-Clause 2 license. 
+// See the license.txt file in the project root for more information.
+
+using System.Collections.Generic;
+using Markdig.Syntax.Inlines;
+
+namespace Markdig.Syntax
+{
+    /// <summary>
+    /// Extensions for visiting <see cref="Block"/> or <see cref="Inline"/>
+    /// </summary>
+    public static class MarkdownObjectExtensions
+    {
+        /// <summary>
+        /// Iterates over the descendant elements for the specified markdown element, including <see cref="Block"/> and <see cref="Inline"/>.
+        /// <para>The descendant elements are returned in DFS-like order.</para>
+        /// </summary>
+        /// <param name="markdownObject">The markdown object.</param>
+        /// <returns>An iteration over the descendant elements</returns>
+        public static IEnumerable<MarkdownObject> Descendants(this MarkdownObject markdownObject)
+        {
+            // Fast-path an object with no children to avoid allocating Stack objects
+            if (!(markdownObject is ContainerBlock) && !(markdownObject is ContainerInline)) yield break;
+
+            // TODO: A single Stack<(MarkdownObject block, bool push)> when ValueTuples are available
+            Stack<MarkdownObject> stack = new Stack<MarkdownObject>();

```

Comment:
You can use instead a small internal struct for that, that's fine

Reply:
Testing with a struct in a single stack I go from a 3.3 to a 4.5 MB increase in allocations. Want me to commit it anyway?